File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderFollowButton.java
package org.wordpress.android.ui.reader.views;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;

import com.google.android.material.button.MaterialButton;

import org.wordpress.android.R;

/**
 * Follow button used in reader detail
 */
public class ReaderFollowButton extends MaterialButton {
    private boolean mIsFollowed;
    private boolean mShowCaption;

    public ReaderFollowButton(Context context) {
        super(context);
        initView(context, null);
    }

    public ReaderFollowButton(Context context, AttributeSet attrs) {
        super(context, attrs, R.attr.followButtonStyle);
        initView(context, attrs);
    }

    public ReaderFollowButton(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initView(context, attrs);
    }

    private void initView(Context context, AttributeSet attrs) {
        // default to showing caption, then read the value from passed attributes
        mShowCaption = true;
        if (attrs != null) {
            TypedArray array = context.getTheme().obtainStyledAttributes(attrs, R.styleable.ReaderFollowButton, 0, 0);
            if (array != null) {
                mShowCaption = array.getBoolean(R.styleable.ReaderFollowButton_wpShowFollowButtonCaption, true);
            }
        }

        if (!mShowCaption) {
            hideCaptionAndEnlargeIcon(context);
        }
    }

    private void hideCaptionAndEnlargeIcon(Context context) {
        setText(null);
        int iconSz = context.getResources().getDimensionPixelSize(R.dimen.reader_follow_icon_no_caption);
        setIconSize(iconSz);
    }

    private void updateFollowTextAndIcon() {
        if (mShowCaption) {
            setText(mIsFollowed ? R.string.reader_btn_unfollow : R.string.reader_btn_follow);
        }
        setSelected(mIsFollowed);

        int drawableId;
        if (mIsFollowed) {
            drawableId = R.drawable.ic_reader_following_white_24dp;
        } else {
            drawableId = R.drawable.ic_reader_follow_white_24dp;
        }
        Drawable drawable = getContext().getResources().getDrawable(drawableId, getContext().getTheme());
        setIcon(drawable);
    }

    public void setIsFollowed(boolean isFollowed) {
        setIsFollowed(isFollowed, false);
    }

    public void setIsFollowedAnimated(boolean isFollowed) {
        setIsFollowed(isFollowed, true);
    }

    private void setIsFollowed(boolean isFollowed, boolean animateChanges) {
        if (isFollowed == mIsFollowed && isSelected() == isFollowed) {
            return;
        }

        mIsFollowed = isFollowed;

        if (animateChanges) {
            ObjectAnimator anim = ObjectAnimator.ofFloat(this, View.SCALE_Y, 1f, 0f);
            anim.setRepeatMode(ValueAnimator.REVERSE);
            anim.setRepeatCount(1);

            anim.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationRepeat(Animator animation) {
                    updateFollowTextAndIcon();
                }
            });

            long duration = getContext().getResources().getInteger(android.R.integer.config_shortAnimTime);
            AnimatorSet set = new AnimatorSet();
            set.play(anim);
            set.setDuration(duration);
            set.setInterpolator(new AccelerateDecelerateInterpolator());

            set.start();
        } else {
            updateFollowTextAndIcon();
        }
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderFollowButton.kt
package org.wordpress.android.ui.reader.views

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import com.google.android.material.button.MaterialButton
import org.wordpress.android.R
import org.wordpress.android.ui.reader.views.ReaderFollowButtonType.FOLLOW_SITE
import android.R as AndroidR

/**
 * Follow button used in reader detail
 */
class ReaderFollowButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.followButtonStyle
) : MaterialButton(context, attrs, defStyleAttr) {
    private var isFollowed = false
    private var showCaption = false
    private var followButtonType = FOLLOW_SITE

    init {
        initView(context, attrs)
    }

    private fun initView(context: Context, attrs: AttributeSet?) {
        // default to showing caption, then read the value from passed attributes
        showCaption = true
        attrs?.let {
            val array = context.theme.obtainStyledAttributes(attrs, R.styleable.ReaderFollowButton, 0, 0)
            showCaption = array.getBoolean(R.styleable.ReaderFollowButton_wpShowFollowButtonCaption, true)

            try {
                val buttonTypeValue = array.getInteger(R.styleable.ReaderFollowButton_wpReaderFollowButtonType, -1)
                if (buttonTypeValue != -1) {
                    followButtonType = ReaderFollowButtonType.fromInt(buttonTypeValue)
                }
            } finally {
                array.recycle()
            }
        }
        if (!showCaption) {
            text = null
        }

        updateFollowText()
    }

    private fun updateFollowText() {
        if (showCaption) {
            setText(if (isFollowed) followButtonType.captionFollowing else followButtonType.captionFollow)
        }
        isSelected = isFollowed
    }

    fun setIsFollowed(isFollowed: Boolean) {
        setIsFollowed(isFollowed, false)
    }

    fun setIsFollowedAnimated(isFollowed: Boolean) {
        setIsFollowed(isFollowed, true)
    }

    @SuppressLint("Recycle")
    private fun setIsFollowed(isFollowed: Boolean, animateChanges: Boolean) {
        if (isFollowed == this.isFollowed && isSelected == isFollowed) {
            return
        }
        this.isFollowed = isFollowed
        if (animateChanges) {
            val anim = ObjectAnimator.ofFloat(
                this,
                View.SCALE_Y,
                1f,
                0f
            )
            with(anim) {
                repeatMode = ValueAnimator.REVERSE
                repeatCount = 1
                addListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationRepeat(animation: Animator) {
                        updateFollowText()
                    }
                })
            }
            AnimatorSet().apply {
                play(anim)
                duration = context.resources.getInteger(AndroidR.integer.config_shortAnimTime).toLong()
                interpolator = AccelerateDecelerateInterpolator()
            }.start()
        } else {
            updateFollowText()
        }
    }
}
